GeneratorInterface/GenFilters/plugins/PythiaFilterMultiAncestor.cc

```diff
@@ -0,0 +1,236 @@
+/*
+This filter allows to select events where a given particle originates *either directly 
+(daughter) or indirectly ((grand)^n-daughter)*  from a list of possible ancestors.
+
+It also allows to filter on the immediate daughters of the said particle.
+
+Kinematic selections can also be applied on the particle and its daughters.
+
+The example below shows how to select a jpsi->mumu coming from *any* b hadron (notice
+that MotherIDs is just 5, i.e. b-quark) however long the decay chain is.
+This includes, for example, B->Jpsi + X as well as B->Psi(2S)(->JPsi) X
+
+process.jpsi_from_bhadron_filter = cms.EDFilter("PythiaFilterMultiAncestor",
+    DaughterIDs = cms.untracked.vint32(-13, 13),
+    DaughterMaxEtas = cms.untracked.vdouble(3., 3.),
+    DaughterMaxPts = cms.untracked.vdouble(100000.0, 100000.0),
+    DaughterMinEtas = cms.untracked.vdouble(-2.6, -2.6),
+    DaughterMinPts = cms.untracked.vdouble(2.5, 2.5),
+    MaxEta = cms.untracked.double(3.0),
+    MinEta = cms.untracked.double(-3.0),
+    MinPt = cms.untracked.double(6.0),
+    MotherIDs = cms.untracked.vint32(5),
+    ParticleID = cms.untracked.int32(443)
+)
+*/
+
+// system include files
+#include <memory>
+#include <iostream>
+
+// user include files
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+
+// #include "FWCore/Framework/interface/EDFilter.h"
+#include "FWCore/Framework/interface/global/EDFilter.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "GeneratorInterface/GenFilters/plugins/MCFilterZboostHelper.h"
```

Comment:
In 10_6_X,  It should be 
#include "GeneratorInterface/GenFilters/interface/MCFilterZboostHelper.h"

Reply:
@qliphy Thanks a lot for pointing this out! 

Looking closer it seems the release now contains two versions of the filter: 
https://github.com/cms-sw/cmssw/blob/dc15e36c8b8c7bbb58963dc9742680add42e1bb5/GeneratorInterface/GenFilters/plugins/PythiaFilterMultiAncestor.cc
https://github.com/cms-sw/cmssw/blob/dc15e36c8b8c7bbb58963dc9742680add42e1bb5/GeneratorInterface/GenFilters/src/PythiaFilterMultiAncestor.cc

Seems like the backport added the version not split into .cc and .h residing inside plugins, whereas the release already contained the filter and parts of the fixes in the src and include folders... 

How should we resolve this? I think it might make sense to split it into .h and .cc in CMSSW 12.3 to mimmic the structure of 10.6.30 and then attempt a backport.

Reply:
> @qliphy Thanks a lot for pointing this out!
> 
> Looking closer it seems the release now contains two versions of the filter: https://github.com/cms-sw/cmssw/blob/dc15e36c8b8c7bbb58963dc9742680add42e1bb5/GeneratorInterface/GenFilters/plugins/PythiaFilterMultiAncestor.cc and https://github.com/cms-sw/cmssw/blob/dc15e36c8b8c7bbb58963dc9742680add42e1bb5/GeneratorInterface/GenFilters/src/PythiaFilterMultiAncestor.cc

You should check the files in either master or 10_6_X branches, and there is no duplication in each:

https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/GenFilters/src/PythiaFilterMultiAncestor.cc
https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/GenFilters/plugins/PythiaFilterMultiAncestor.cc

https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/GeneratorInterface/GenFilters/src/PythiaFilterMultiAncestor.cc
https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/GeneratorInterface/GenFilters/plugins/PythiaFilterMultiAncestor.cc

> How should we resolve this? I think it might make sense to split it into .h and .cc in CMSSW 12.3 to mimmic the structure of 10.6.30 and then attempt a backport.

10_6_X and master are not necessary to be the same in this matter. You don't need to split the file into .h and .cc.

Reply:
@qliphy, In CMSSW 12_3_X, the Genfilters directory structure differs from CMSSW 10_6_X though. Does the backport bound by the directory structure for consistency?

Reply:
> @qliphy, In CMSSW 12_3_X, the Genfilters directory structure differs from CMSSW 10_6_X though. Does the backport bound by the directory structure for consistency?

I don't think so.  One can keep the 10_6_X convention.   
I just notice that there are already 10_6_X backports:  
                    https://github.com/cms-sw/cmssw/pull/32399
                    https://github.com/cms-sw/cmssw/pull/32450
So now you only need to backport 
                    https://github.com/cms-sw/cmssw/pull/33491

Reply:
I made another PR with just the last fix: https://github.com/cms-sw/cmssw/pull/37065 I guess we could close this PR.

Reply:
@mhuwiler  Thanks.  Yes this one can be closed.